url04f7fcf4e1_peteratt_HDFS_Replication_Simulator_tgz-pJ8-hdfs_replicationsimulator_MainJ8/src/hdfs/replicationsimulator/SimTrace.java

```diff
@@ -0,0 +1,104 @@
+package hdfs.replicationsimulator;
+
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+
+public class SimTrace {
+	
+	public static final int FAILURE_DETECTION = 0;
+	public static final int SCHEDULED_FOR_REPAIR = 1;
+	public static final int DATA_LOSS = 2;
+	public static final int BLOCK_RECEIVED = 3;
+	public static final int INFO = 4;
+	
+	private long timestamp;
+	private String message;
+	
+	private int action;
+	private int id;
+	private int idBlock;
+	
+	public SimTrace(long timestamp, String msg) {
+		this.timestamp = timestamp;
+		this.message = msg;
+	}
+	
+	public SimTrace(int action, int id) {
+		this.timestamp = Node.now();
+		this.action = action;
+		this.id = id;
+		
+		switch (action) {
+		
+		case FAILURE_DETECTION:
+			this.message = "Namenode has detected a failure in Datanode #" + id;
+			break;
+			
+		case SCHEDULED_FOR_REPAIR:
+			this.message = "Namenode has scheduled to repair block #" + id;
+			break;
+			
+		case DATA_LOSS:
+			this.message = "Block #" + id + " is lost";
+			break;
+		}
+	}
+	
+	public SimTrace(String info) {
+		this.timestamp = Node.now();
+		this.action = INFO;
+		this.id = 0;
+		this.message = info;
+	}
+	
+	public SimTrace(int action, int id, int idBlock) {
+		this.timestamp = Node.now();
+		this.action = action;
+		this.id = id;
+		this.idBlock = idBlock;
+		
+		switch (action) {
+			
+		case BLOCK_RECEIVED:
+			this.message = "Datanode #" + id + " has received a block #" + idBlock;
+			break;
+		}
+	}
+	
+	public long getTimestamp() {
+		return timestamp;
+	}
+	
+	public String getMessage() { 
+		return message;
+	}
+	
+	@Override
+	public String toString() {
+		
+		Calendar calendar = Calendar.getInstance();
+		calendar.setTimeInMillis(timestamp);
+		
+		DateFormat formatter = new SimpleDateFormat("dd/MM/yyyy hh:mm:ss.SSS");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The SimpleDateFormat object was created without setting the time zone. Make sure that you want to use the default time zone or use letter z, Z or X in the pattern. Otherwise, if you do not call setTimeZone() on the created SimpleDateFormat object, you might get an unexpected date and time when using the object.